contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,434 @@
+4. Upon successful completion of the `ControllerPublish` call the external-attacher updates the status of the `VolumeAttachment` object to indicate the volume is successful attached.
+5. The in-tree volume plugin watching the status of the `VolumeAttachment` object in the kubernetes API, sees the `IsAttached` field set to true indicating the volume is attached, so It updates the attach/detach controller’s internal state to indicate the volume is attached.
+
+#### Detaching Volumes
+
+1. The Kubernetes attach/detach controller, running as part of the `kube-controller-manager` binary on the master, sees that a pod referencing an attached CSI volume plugin is terminated or deleted, so it calls the in-tree CSI volume plugin’s detach method.
+2. The in-tree volume plugin deletes the corresponding `VolumeAttachment` object.
+3. The external-attacher sees a `deletionTimestamp` set on the `VolumeAttachment` object and triggers a `ControllerUnpublish` against the CSI volume driver container to detach it.
+4. Upon successful completion of the `ControllerUnpublish` call, the external-attacher removes the finalizer from the `VolumeAttachment` object to indicate successful completion of the detach operation allowing the `VolumeAttachment` object to be deleted.
+5. The in-tree volume plugin waiting for the `VolumeAttachment` object sees it deleted and assumes the volume was successfully detached, so It updates the attach/detach controller’s internal state to indicate the volume is detached.
+
+#### Mounting Volumes
+
+1. The volume manager component of kubelet notices a new volume, referencing a CSI volume, has been scheduled to the node, so it calls the in-tree CSI volume plugin’s `WaitForAttach` method.
+2. The in-tree volume plugin’s `WaitForAttach` method watches the `IsAttached` field of the `VolumeAttachment` object in the kubernetes API to become `true`, it then returns without error.
+3. Kubelet then calls the in-tree CSI volume plugin’s `MountDevice` method which is a no-op and returns immediately.
+4. Finally kubelet calls the in-tree CSI volume plugin’s mount (setup) method, which causes the in-tree volume plugin to issue a `NodePublishVolume` call via the registered unix domain socket to the local CSI driver.
+5. Upon successful completion of the `NodePublishVolume` call the specified path is mounted into the pod container.
+
+#### Unmounting Volumes
+1. The volume manager component of kubelet, notices a mounted CSI volume, referenced by a pod that has been deleted or terminated, so it calls the in-tree CSI volume plugin’s `UnmountDevice` method which is a no-op and returns immediately.
+2. Next kubelet calls the in-tree CSI volume plugin’s unmount (teardown) method, which causes the in-tree volume plugin to issue a `NodeUnpublishVolume` call via the registered unix domain socket to the local CSI driver.
+3. Upon successful completion of the `NodeUnpublishVolume` call the specified path is mounted into the pod container.
```

Comment:
Nit, _unmounted_ from the container, but please note what happens if the call fails for any reason. The volume is left marked as mounted on the node.

Reply:
> unmounted from the pod container

fixed

> what happens if the call fails for any reason. The volume is left marked as mounted on the node

Kubelet tries to un-mount the volume periodically already today, we don't plan to change this behavior.